### PR TITLE
feat: Save `am instrument` logs to file & handle parsing error.

### LINKED ISSUE #2037: Handle 'am instrument' errors
## Motivation

Currently, Flank can properly parse `am instrument` output only when it contains results from running tests.
Any other cases are not handled so the error messages can be confusing.

## Goal

Handle properly different types of output from `am instrument` command to improve overall UX.

## Example problem

For example, running APK for x86 architecture on ARM Corellium device will result in the following error, because the flank is trying to parse some unknown output as test output:

```
java.util.NoSuchElementException: Key current is missing in the map.
	at kotlin.collections.MapsKt__MapWithDefaultKt.getOrImplicitDefaultNullable(MapWithDefault.kt:24)
	at kotlin.collections.MapsKt__MapsKt.getValue(Maps.kt:344)
	at flank.instrument.log.InternalKt.getId(Internal.kt:132)
	at flank.instrument.log.InternalKt.access$getId(Internal.kt:1)
	at flank.instrument.log.InternalKt$parseStatusResult$$inlined$transform$1$1.emit(Collect.kt:140)
	at flank.instrument.log.InternalKt$parseChunks$$inlined$map$1$2.emit(Collect.kt:157)
	at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt:15)
	at kotlinx.coroutines.flow.internal.SafeCollectorKt$emitFun$1.invoke(SafeCollector.kt:15)
	at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:77)
	at kotlinx.coroutines.flow.internal.SafeCollector.emit(SafeCollector.kt:59)
	at flank.instrument.log.InternalKt$groupLines$$inlined$transform$1$1.emit(Collect.kt:141)
	at kotlinx.coroutines.flow.FlowKt__LimitKt$dropWhile$lambda-4$$inlined$collect$1.emit(Collect.kt:136)
	at kotlinx.coroutines.flow.FlowKt__ChannelsKt.emitAllImpl$FlowKt__ChannelsKt(Channels.kt:62)
	at kotlinx.coroutines.flow.FlowKt__ChannelsKt.access$emitAllImpl(Channels.kt:1)
	at kotlinx.coroutines.flow.FlowKt__ChannelsKt$emitAllImpl$1.invokeSuspend(Channels.kt)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at flank.corellium.domain.RunTestCorelliumAndroidKt.invoke(RunTestCorelliumAndroid.kt:137)
	at flank.corellium.cli.RunTestCorelliumAndroidCommand.run(RunTestCorelliumAndroidCommand.kt:143)
	at picocli.CommandLine.executeUserObject(CommandLine.java:1939)
	at picocli.CommandLine.access$1300(CommandLine.java:145)
	at picocli.CommandLine$RunLast.executeUserObjectOfLastSubcommandWithSameParent(CommandLine.java:2352)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2346)
	at picocli.CommandLine$RunLast.handle(CommandLine.java:2311)
	at picocli.CommandLine$AbstractParseResultHandler.execute(CommandLine.java:2179)
	at picocli.CommandLine.execute(CommandLine.java:2078)
	at ftl.Main$main$1.invoke(Main.kt:12)
	at ftl.Main$main$1.invoke(Main.kt:10)
	at ftl.run.exception.ExceptionHandlerKt.withGlobalExceptionHandling(ExceptionHandler.kt:28)
	at ftl.run.exception.ExceptionHandlerKt.withGlobalExceptionHandling(ExceptionHandler.kt:17)
	at ftl.Main.main(Main.kt:10)

```

### PR DESCRIPTION
**Pull request by jan-gogo**
Fixes #2037 

Improvements:

* Save `am instrument` logs to file in `adb_log` output directory.
* Catch `am instrument` output parsing error and mark affected lines in the error message.
* Close console connection without waiting for idle, if receive expected results or error.

* Unit test passes. 
* Running `flank corellium test android run` creates `adb_log` inside output directory with dumped `am instrument` console outputs.
* If parsing tests results from `am instrument` log fails the error message in the console will communicate file with lines where the error occurs.

## Checklist

- [x] Documented
- [x] Unit tested

### CODE REVIEW

File: corellium/domain/src/test/kotlin/flank/corellium/domain/run/test/android/step/ExecuteTestsKtTest.kt [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    override var out: Output = ::println
+    override val apk = Apk.Api()
+    override val junit = JUnit.Api()
+    override val args = RunTestCorelliumAndroid.Args(
+        credentials = stubCredentials,
+        apks = emptyList(),
+        maxShardsCount = 1,
+    )
+
+    private val dir = File(args.outputDir)
```
Comment (pawelpasterz) [inline corellium/domain/src/test/kotlin/flank/corellium/domain/run/test/android/step/ExecuteTestsKtTest.kt:40]: Instead of `File` use `TemporaryFolder` (JUnit). Current implementation does not guarantee clean up after failed tests/exception thrown

File: corellium/adapter/src/test/kotlin/flank/corellium/adapter/ExecuteAndroidTestPlanKtTest.kt [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.channelFlow
+import kotlinx.coroutines.flow.first
+import kotlinx.coroutines.runBlocking
+import org.junit.Assert
+import org.junit.Test
+
+class ExecuteAndroidTestPlanKtTest {
```
Comment (pawelpasterz) [inline corellium/adapter/src/test/kotlin/flank/corellium/adapter/ExecuteAndroidTestPlanKtTest.kt:24]: Please add `unmockkAll()` tear down method -- unmocked tests may interact between test suites (we had this issue couple of times in the past)

File: corellium/cli/src/main/kotlin/flank/corellium/cli/RunTestCorelliumAndroidCommand.kt [hunk 3]
```diff
@@ -232,6 +232,13 @@ internal val format = buildFormatter<String> {
             else -> null
         }
     }
+    ExecuteTests.Error::class {
+        """
+            Error while parsing results from instance $id.
+            For details check $logFile lines $lines.
+
```
Comment (pawelpasterz) [inline corellium/cli/src/main/kotlin/flank/corellium/cli/RunTestCorelliumAndroidCommand.kt:239]: Blank line left intentionally?

**Review by pawelpasterz — CHANGES_REQUESTED** [review]

File: corellium/domain/src/test/kotlin/flank/corellium/domain/run/test/android/step/ExecuteTestsKtTest.kt (hunk 1 above)
Comment (jan-gogo) [reply corellium/domain/src/test/kotlin/flank/corellium/domain/run/test/android/step/ExecuteTestsKtTest.kt:40]: > Instead of `File` use `TemporaryFolder` (JUnit). Current implementation does not guarantee clean up after failed tests/exception thrown

I do not agree. This is the first shot from duckduckgo for the phase `JUnit After`:  https://junit.org/junit4/javadoc/4.12/org/junit/After.html

File: corellium/cli/src/main/kotlin/flank/corellium/cli/RunTestCorelliumAndroidCommand.kt (hunk 3 above)
Comment (jan-gogo) [reply corellium/cli/src/main/kotlin/flank/corellium/cli/RunTestCorelliumAndroidCommand.kt:239]: Yup

File: corellium/adapter/src/test/kotlin/flank/corellium/adapter/ExecuteAndroidTestPlanKtTest.kt (hunk 2 above)
Comment (jan-gogo) [reply corellium/adapter/src/test/kotlin/flank/corellium/adapter/ExecuteAndroidTestPlanKtTest.kt:26]: Fixed

File: corellium/domain/src/test/kotlin/flank/corellium/domain/run/test/android/step/ExecuteTestsKtTest.kt (hunk 1 above)
Comment (pawelpasterz) [reply corellium/domain/src/test/kotlin/flank/corellium/domain/run/test/android/step/ExecuteTestsKtTest.kt:40]: And this is the short from google (I do not use duckduckgo, maybe I should start to do so)
https://junit.org/junit4/javadoc/4.12/org/junit/rules/TemporaryFolder.html
 
`@After` may throw an exception itself which will break i.e. recursive deletion. 
Not saying your implementation will not work, most likely it will be working fine.
`TemporaryFolder` is implemented to be used in exactly the same cases we have in this particular test.

**Review by pawelpasterz — COMMENTED** [review]
I'm not giving the approval since I do not agree with some decisions in tests (personal preferences). But I don't want to block anything when it should work fine.

File: corellium/domain/src/test/kotlin/flank/corellium/domain/run/test/android/step/ExecuteTestsKtTest.kt (hunk 1 above)
Comment (jan-gogo) [reply corellium/domain/src/test/kotlin/flank/corellium/domain/run/test/android/step/ExecuteTestsKtTest.kt:40]: Ok, so according to documentation, the `TemporaryFolder` is also not guaranteed clean up same as `@After`. Additionally, it will not throw an exception on recursive deletion fail. This is bad because the test should always clean up its own garbages, if it is not possible that means probably some external process is reading the files created during the test. This situation is an edge case that should throw an exception.

**Review by pawelpasterz — APPROVED** [review]
> I'm not giving the approval since I do not agree with some decisions in tests (personal preferences). But I don't want to block anything when it should work fine.

ah, whatever

**Comment by jan-gogo** [comment]
> ah, whatever

Thx 😄 👍

**Review by Sloox — APPROVED** [review]
